File: src/app/model/PowerMetric.ts
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";
import { PowerMeterServiceEvent } from "./PowerMeterServiceEvent";

export class PowerMetric extends Metric {
    private handler = this.handlePowerMeterEvent.bind(this);

    constructor(metricService: MetricService) {
        super(MetricType.POWER, 'Power', 'W', metricService);
    }

    startLogging(): void {
        this.metricService.getPowerMeterService().subscribeForPowerData(this.handler);
    }


    stopLogging(): void {
        this.metricService.getPowerMeterService().unsubscribeForPowerData(this.handler);
    }

    handlePowerMeterEvent(event: PowerMeterServiceEvent): void {
        let power = event.powermeterData.power;
        if (event.powermeterData.balence === null) {
            // If balence is null, we assume the power is from the left side only
            power = power * 2; // Double the power to account for both sides
        }
        this.addValue(power, new Date(event.powermeterData.timestamp));
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/PowerMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnPowerData;

public class PowerMetric extends Metric implements OnPowerData {
    public PowerMetric(IMetricService metricService) {
        super(MetricType.POWER, "Power", "W", metricService);
    }

    @Override
    public void startLogging() {
        metricService.getPowerMeterService().subscribeForPowerData(this);
    }

    @Override
    public void stopLogging() {
        metricService.getPowerMeterService().unsubscribeForPowerData(this);
    }

    @Override
    public void onPowerData(PowerMeterData data) {
        addValue(data.getTotalPower(), data.timestamp);
    }
}
